File: cpp/lib/testcontainers/core/wait/WaitFor.cpp
#include <rust/cxx.h>
#include <rust_tc_bridge/lib.h>

#include "testcontainers/core/wait/WaitFor.hpp"

#include "details/BoxHelper.hpp"

namespace testcontainers {

WaitFor::WaitFor(RsWaitFor *wait_for) noexcept
    : rimpl_(wait_for, [](RsWaitFor *w) { ::rs_wait_for_destroy(details::box_from_raw(w)); }) {}

WaitFor::WaitFor(const WaitFor &other) : WaitFor(other.rimpl_->rs_wait_for_clone().into_raw()) {}

WaitFor::WaitFor(WaitFor &&other) noexcept = default;

WaitFor &WaitFor::operator=(WaitFor &&other) noexcept = default;

WaitFor::~WaitFor() noexcept = default;

WaitFor WaitFor::Nothing() noexcept { return WaitFor(::rs_wait_for_nothing().into_raw()); }

WaitFor WaitFor::Log(LogWaitStrategy strategy) noexcept {
  return WaitFor(::rs_wait_for_log(details::into_box(strategy.rimpl_)).into_raw());
}

WaitFor WaitFor::Duration(std::chrono::duration<std::uint64_t, std::nano> duration) noexcept {
  return WaitFor(::rs_wait_for_duration(duration.count()).into_raw());
}

WaitFor WaitFor::Healthcheck(HealthWaitStrategy strategy) noexcept {
  return WaitFor(::rs_wait_for_healthcheck(details::into_box(strategy.rimpl_)).into_raw());
}

WaitFor WaitFor::Exit(ExitWaitStrategy strategy) noexcept {
  return WaitFor(::rs_wait_for_exit(details::into_box(strategy.rimpl_)).into_raw());
}

WaitFor WaitFor::message_on_stdout(std::string_view message) {
  return Log(LogWaitStrategy::std_out(message));
}

WaitFor WaitFor::message_on_stderr(std::string_view message) {
  return Log(LogWaitStrategy::std_err(message));
}

WaitFor WaitFor::message_on_either_std(std::string_view message) {
  return Log(LogWaitStrategy::stdout_or_stderr(message));
}

WaitFor WaitFor::log(LogWaitStrategy strategy) noexcept {
  return WaitFor::Log(std::move(strategy));
}

WaitFor WaitFor::healthcheck() noexcept {
  return WaitFor::Healthcheck(HealthWaitStrategy::healthcheck());
}

WaitFor WaitFor::exit() noexcept { return WaitFor::Exit(ExitWaitStrategy::exit()); }

WaitFor WaitFor::seconds(std::uint64_t seconds) noexcept {
  return WaitFor::Duration(std::chrono::seconds(seconds));
}

WaitFor WaitFor::millis(std::uint64_t milliseconds) noexcept {
  return WaitFor::Duration(std::chrono::milliseconds(milliseconds));
}

WaitFor WaitFor::millis_in_env_var(std::string_view name) {
  return WaitFor(::rs_wait_for_millis_in_env_var(details::into_str(name)).into_raw());
}

bool WaitFor::is_valid() const noexcept { return static_cast<bool>(rimpl_); }

} // namespace testcontainers

File: rust/src/core/wait/wait_for.rs
use crate::core::wait::{
    exit_wait_strategy::RsExitWaitStrategy, health_wait_strategy::RsHealthWaitStrategy,
    log_wait_strategy::RsLogWaitStrategy,
};
use std::time::Duration;
use testcontainers::core::WaitFor;

pub struct RsWaitFor {
    pub strategy: WaitFor,
}

pub fn rs_wait_for_nothing() -> Box<RsWaitFor> {
    Box::new(RsWaitFor::new(WaitFor::Nothing))
}

pub fn rs_wait_for_log(strategy: Box<RsLogWaitStrategy>) -> Box<RsWaitFor> {
    Box::new(RsWaitFor::new(WaitFor::Log(strategy.strategy)))
}

pub fn rs_wait_for_duration(duration_ns: u64) -> Box<RsWaitFor> {
    Box::new(RsWaitFor::new(WaitFor::Duration {
        length: Duration::from_nanos(duration_ns),
    }))
}

pub fn rs_wait_for_healthcheck(strategy: Box<RsHealthWaitStrategy>) -> Box<RsWaitFor> {
    Box::new(RsWaitFor::new(WaitFor::Healthcheck(strategy.strategy)))
}

pub fn rs_wait_for_exit(strategy: Box<RsExitWaitStrategy>) -> Box<RsWaitFor> {
    Box::new(RsWaitFor::new(WaitFor::Exit(strategy.strategy)))
}

pub fn rs_wait_for_millis_in_env_var(name: &'static str) -> Box<RsWaitFor> {
    Box::new(RsWaitFor::new(WaitFor::millis_in_env_var(name)))
}

// vec helpers
pub fn rs_wait_for_vec_push(vec: &mut Vec<RsWaitFor>, wait_for: Box<RsWaitFor>) {
    vec.push(*wait_for);
}

pub fn rs_wait_for_vec_pop(vec: &mut Vec<RsWaitFor>) -> Result<Box<RsWaitFor>, String> {
    vec.pop()
        .map(Box::new)
        .ok_or_else(|| "Vector is empty".to_string())
}

pub fn rs_wait_for_destroy(wait_for: Box<RsWaitFor>) {
    drop(wait_for);
}

impl RsWaitFor {
    pub fn new(strategy: WaitFor) -> Self {
        Self { strategy }
    }

    pub fn rs_wait_for_clone(self: &RsWaitFor) -> Box<RsWaitFor> {
        Box::new(RsWaitFor::new(self.strategy.clone()))
    }
}
